WPB-21964: Add Wire Meetings endpoints

blackheaven opened the pull request:
https://wearezeta.atlassian.net/browse/WPB-21964

## Checklist

 - [x] Add a new entry in an appropriate subdirectory of `changelog.d`
 - [x] Read and follow the [PR guidelines](https://docs.wire.com/latest/developer/developer/pr-guidelines.html)

blackheaven committed:
WPB-21964: Add Wire Meetings endpoints

blackheaven commented on libs/wire-subsystems/src/Wire/MeetingsSubsystem/Interpreter.hs:
```diff
@@ -0,0 +1,323 @@
+
+  -- Determine trial status
+  -- TODO: Check if user is a paying customer via Feature
+  let trial = False
```
Do you know which feature flag should be checked?

akshaymankar commented on libs/wire-api/src/Wire/API/Meeting.hs:
```diff
@@ -0,0 +1,133 @@
+    creator :: UserId,
+    startDate :: UTCTime,
+    endDate :: UTCTime,
+    schedule :: Maybe Text,
```
IMO the cron format is a mistake here, there is some discussion about it in the RFC. We should wait for that to resolve.

akshaymankar reviewed (changes requested):
There are more questions about this in the RFC. Let's wait before implementation.

blackheaven committed:
fix: rework feature flags (meeting, meeting_premium)

battermann commented on integration/test/Test/FeatureFlags/PayingTeam.hs:
```diff
@@ -0,0 +1,31 @@
+-- You should have received a copy of the GNU Affero General Public License along
+-- with this program. If not, see <https://www.gnu.org/licenses/>.
+
+module Test.FeatureFlags.PayingTeam where
```
this should be renamed as well

battermann commented on integration/integration.cabal:
```diff
@@ -157,6 +157,7 @@ library
     Test.FeatureFlags.MlsE2EId
     Test.FeatureFlags.MlsMigration
     Test.FeatureFlags.OutlookCalIntegration
+    Test.FeatureFlags.PayingTeam
```
rename

battermann commented on libs/wire-api/src/Wire/API/Routes/Internal/Galley.hs:
```diff
@@ -94,6 +94,8 @@ type IFeatureAPI =
     :<|> IFeatureStatusLockStatusPut AppsConfig
     :<|> IFeatureStatusLockStatusPut SimplifiedUserConnectionRequestQRCodeConfig
     :<|> IFeatureStatusLockStatusPut StealthUsersConfig
+    :<|> IFeatureStatusLockStatusPut MeetingConfig
```
if meetings is unlocked by default maybe we should consider to not enable the modification of the lock status at all because there is no requirement from the business perspective. (why would you ever want to lock it)

battermann commented on libs/wire-api/src/Wire/API/Meeting.hs:
```diff
@@ -0,0 +1,163 @@
+newtype MeetingId = MeetingId {unMeetingId :: UUID}
+  deriving stock (Eq, Ord, Show, Generic)
+  deriving newtype (ToJSON, FromJSON, FromHttpApiData, ToHttpApiData, S.ToSchema, S.ToParamSchema)
+  deriving (Arbitrary) via (GenericUniform MeetingId)
```
consider putting this into `Data.Id`

battermann commented on services/galley/src/Galley/API/Meetings.hs:
```diff
@@ -0,0 +1,235 @@
+            _ -> True -- non-premium team or disabled, is trial
+  (meeting, conversation) <- Meetings.createMeeting lUser newMeeting trial
+  notifyCreatedConversation lUser Nothing conversation InternalAdd
+  pure meeting
```
I know, the current state does not suggest this, but we want all the business logic to be implemented in the subsystems. The code in the services should only be responsible for bootstrapping and wiring up the API. So all of this should go into the MeetingSubsystemInterpreter.

battermann commented on services/galley/postgresql-migrations/001_meetings.sql:
```diff
@@ -0,0 +1,102 @@
+    -- Scheduling information
+    start_date timestamptz NOT NULL,
+    end_date timestamptz NOT NULL,
+    recurrence jsonb,
```
Not sure if we should store the recurrence untyped, like this. This is really an open question, not a critique.

battermann reviewed (changes requested):
For the feature flags we are missing:
- stern endpoints and tests
- helm charts, configmap, default values, docs etc.

For the other config (background-worker) we are missing the helm charts etc. too

(see PR guidelines)

blackheaven committed:
feat(stern): Add Meeting and MeetingPremium feature flags to Stern API

Adds routes for MeetingConfig and MeetingPremiumConfig to the Stern API,
including get, put, and lock/unlock routes.

Updates the galley helm chart values to include default configurations
for Meeting and MeetingPremium feature flags, and ensures these are
properly exposed through the ConfigMap.

Addresses part of the review comment regarding missing Stern endpoints,
helm charts, and background worker config.

blackheaven committed:
feat(meetings): Replace payingTeam with meeting and meetingPremium feature flags

The `payingTeam` feature flag is replaced by two new feature flags:

- `meeting`: Controls access to the meetings API. When disabled, all meetings
  endpoints return 403 Forbidden. The feature is enabled and unlocked by
  default.
- `meetingPremium`: Indicates whether a team has premium meeting features.
  When enabled, meetings created by team members are not marked as trial.

This change also includes:
- Renaming the `PayingTeam` feature flag test to `Meeting`.
- Adding a new `MeetingPremium` feature flag test.
- Updating helm charts and integration configurations.

blackheaven committed:
fix: schema recurrence field

blackheaven committed:
docs(api): frequency doc

blackheaven committed:
docs: Add documentation for meeting and meetingPremium feature flags

Adds documentation for the `meeting` and `meetingPremium` feature flags
to `docs/src/developer/reference/config-options.md`.

blackheaven committed:
refactor: move MeetingId to Data.Id

blackheaven committed:
feat: Add updated_at column to meetings table and update relevant functions

blackheaven committed:
fix: list test name

blackheaven committed:
fix: conversation creatio permissions

blackheaven committed:
feat: Add recurrence tests and update existing tests to handle recurrence

blackheaven committed:
feat: Add date validation tests for meeting creation and updates

blackheaven committed:
refactor: Move meetings migration to wire-subsystems

blackheaven committed:
fix(meetings): Correct JSON paths in integration tests

The previous implementation of the meetings integration tests used incorrect JSON paths to access fields in the response bodies. This resulted in test failures.

This commit corrects the following issues:
- Uses 'qualified_id' to access the 'id' and 'domain' of a meeting.
- Uses 'qualified_creator' to access the creator's ID.
- Corrects the JSON field name for invited emails from 'invitedEmails' to 'invited_emails'.
- Fixes the recurrence frequency field name from 'freq' to 'frequency'.
- Corrects the extraction of 'teamId' from the 'createTeam' function's return value.

blackheaven committed:
feat(meetings): Use lowercase for frequency serialization

This commit updates the 'Frequency' enum to serialize to and from
lowercase strings (e.g., "daily", "weekly").

- The  instance now uses  to convert the constructor
  names to lowercase.
- The  instance is updated to handle both uppercase and
  lowercase values for backward compatibility.
- The  instance is updated to reflect the new lowercase values.
- The integration tests are updated to use and expect lowercase
  frequency values.

blackheaven committed:
refactor: apply suggestions in libs/wire-subsystems/src/Wire/MeetingsStore/Postgres.hs

blackheaven committed:
refactor: clean integration test

blackheaven committed:
fix(cleanup): ensure conversation is associated

blackheaven committed:
fix: listMeetingsImpl returns also meeting the user is a simple participants

blackheaven commented on libs/wire-subsystems/src/Wire/MeetingsSubsystem/Interpreter.hs:
```diff
@@ -0,0 +1,318 @@
+    newMeeting.recurrence
+    qConvId
+    newMeeting.invitedEmails
+    trial
```
I have moved a part in the interpreter, but feature flags are too coupled with galley

blackheaven committed:
refactor: split clean up and regular meeting subsystem, partially move api logic inn the subsystem

blackheaven committed:
fix: patching meeting dates

blackheaven committed:
refactor(tests): use common status assertions

blackheaven committed:
refactor: drop multi-qualified

blackheaven committed:
fix: formatting

blackheaven committed:
fix: hlint

blackheaven committed:
feature(background-worker): use cron

blackheaven committed:
fix: queries exclude expired meetings

blackheaven committed:
fix: formatting
